cmake 3.5.2 build issue

OS: `Slackware6-14.2`
cmake: `3.5.2`
gfxconstruct: https://github.com/LunarG/gfxreconstruct/commit/45bf3b5999d9b590a671e7274cfa53e4f251a25d

In the `cmake/CodeStyle.cmake` file gfxconstruct requires the minimum cmake version of `3.4.1`.

https://github.com/LunarG/gfxreconstruct/blob/45bf3b5999d9b590a671e7274cfa53e4f251a25d/cmake/CodeStyle.cmake#L20

And then uses `FILTER` for the `list` command.

https://github.com/LunarG/gfxreconstruct/blob/45bf3b5999d9b590a671e7274cfa53e4f251a25d/cmake/CodeStyle.cmake#L44

However `FILTER` was not available until `3.6`.

https://cmake.org/cmake/help/v3.6/command/list.html

And is not available for `3.4` which will result in a build issue.

https://cmake.org/cmake/help/v3.4/command/list.html

I suggest this can be solved in one of two ways.

* Require a higher cmake version, many projects associated with the `vulkan-sdk` require at least `3.10.2`.
* Fix the cmake macro to not use `FILTER`.

Removing the single use of the macro seems enough to work around the issue in my case.

https://github.com/LunarG/gfxreconstruct/blob/45bf3b5999d9b590a671e7274cfa53e4f251a25d/cmake/CodeStyle.cmake#L49

Build error:
```
-- The C compiler identification is GNU 5.5.0
-- The CXX compiler identification is GNU 5.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Vulkan Header = /usr/include/vulkan
-- Detected Vulkan Version 0.0.0
-- Could NOT find LZ4 (missing:  LZ4_LIBRARY LZ4_INCLUDE_DIR) 
-- Could NOT find ZSTD (missing:  ZSTD_LIBRARY ZSTD_INCLUDE_DIR) 
-- Found XCB: /usr/lib64/libxcb.so;/usr/lib64/libxcb-keysyms.so;/usr/lib64/libX11-xcb.so  
-- Could NOT find WAYLAND (missing:  WAYLAND_LIBRARY WAYLAND_INCLUDE_DIR) 
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.11") 
-- Looking for XOpenDisplay in /usr/lib64/libX11.so;/usr/lib64/libXext.so
-- Looking for XOpenDisplay in /usr/lib64/libX11.so;/usr/lib64/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- Found X11: /usr/lib64/libX11.so
-- Looking for clock_gettime in c
-- Looking for clock_gettime in c - found
CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  framework/util/CMakeLists.txt:120 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  framework/format/CMakeLists.txt:40 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  framework/encode/CMakeLists.txt:74 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  framework/decode/CMakeLists.txt:117 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  framework/application/CMakeLists.txt:55 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  layer/CMakeLists.txt:41 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  tools/replay/CMakeLists.txt:35 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  tools/toascii/CMakeLists.txt:34 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  tools/compress/CMakeLists.txt:36 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  tools/info/CMakeLists.txt:34 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  tools/extract/CMakeLists.txt:34 (common_build_directives)


CMake Error at cmake/CodeStyle.cmake:44 (list):
  list does not recognize sub-command FILTER
Call Stack (most recent call first):
  cmake/CodeStyle.cmake:49 (generate_target_source_files)
  CMakeLists.txt:96 (target_code_style_build_directives)
  tools/optimize/CMakeLists.txt:35 (common_build_directives)


-- Configuring incomplete, errors occurred!
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -83,7 +83,7 @@ endif()
 
 configure_file("${CMAKE_SOURCE_DIR}/project_version.h.in" "${CMAKE_BINARY_DIR}/project_version.h")
 
-option(BUILD_WERROR "Build with -Werror" ON)
+option(BUILD_WERROR "Build with warnings as errors" ON)
 
 # Code checks
 include("CodeStyle")
@@ -126,7 +126,6 @@ endif(UNIX)
 
 add_library(windows_specific INTERFACE)
 target_compile_definitions(windows_specific INTERFACE WIN32_LEAN_AND_MEAN NOMINMAX VK_USE_PLATFORM_WIN32_KHR)
-target_compile_options(windows_specific INTERFACE /WX)
 
 add_library(linux_specific INTERFACE)
 target_compile_definitions(linux_specific INTERFACE _FILE_OFFSET_BITS=64 PAGE_GUARD_ENABLE_UCONTEXT_WRITE_DETECTION
@@ -135,14 +134,18 @@ target_compile_definitions(linux_specific INTERFACE _FILE_OFFSET_BITS=64 PAGE_GU
     $<$<BOOL:${XCB_FOUND}>:VK_USE_PLATFORM_XCB_KHR>
     $<$<BOOL:${WAYLAND_FOUND}>:VK_USE_PLATFORM_WAYLAND_KHR>)
 
-if(BUILD_WERROR)
-    target_compile_options(linux_specific INTERFACE -Werror)
-endif(BUILD_WERROR)
-
 add_library(platform_specific INTERFACE)
 target_link_libraries(platform_specific INTERFACE
     $<$<BOOL:${WIN32}>:windows_specific>
     $<$<NOT:$<BOOL:${WIN32}>>:linux_specific>)
+
+if(BUILD_WERROR)
+    if(MSVC)
+        target_compile_options(platform_specific INTERFACE /WX)
+    else()
+        target_compile_options(platform_specific INTERFACE -Werror)
+    endif()
+endif()
 
 add_library(vulkan_registry INTERFACE)
 target_include_directories(vulkan_registry INTERFACE ${CMAKE_SOURCE_DIR}/external/Vulkan-Headers/include)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -20,7 +20,7 @@
 # Description: CMake script for framework util target
 ###############################################################################
 
-cmake_minimum_required(VERSION 3.4.1)
+cmake_minimum_required(VERSION 3.10.2)
 project(GFXReconstruct)
 
 set_property(GLOBAL PROPERTY TEST_SCRIPT ${CMAKE_ROOT})

--- a/cmake/CodeStyle.cmake
+++ b/cmake/CodeStyle.cmake
@@ -44,27 +44,29 @@ endmacro()
 
 # Apply code style build directives
 function(target_code_style_build_directives TARGET)
-    generate_target_source_files(${TARGET} TARGET_SRC_FILES TARGET_SOURCE_FILES)
-    if(${APPLY_CPP_CODE_STYLE})
-        if(CLANG_FORMAT-NOTFOUND STREQUAL ${CLANG_FORMAT})
-            message(FATAL_ERROR "Failed to find clang-format in system path")
+    if(${APPLY_CPP_CODE_STYLE} OR ${CHECK_CPP_CODE_STYLE})
+        generate_target_source_files(${TARGET} TARGET_SRC_FILES TARGET_SOURCE_FILES)
+        if(${APPLY_CPP_CODE_STYLE})
+            if(CLANG_FORMAT-NOTFOUND STREQUAL ${CLANG_FORMAT})
+                message(FATAL_ERROR "Failed to find clang-format in system path")
+            endif()
+            # If apply code style is on, turn off check code style
+            set(CHECK_CPP_CODE_STYLE OFF)
+            add_custom_target("${TARGET}ClangFormat"
+                    COMMAND ${CLANG_FORMAT} -i ${TARGET_SRC_FILES}
+                    WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}
+                    COMMENT "Run clang format for ${TARGET}"
+                    COMMAND_EXPAND_LISTS)
+            add_dependencies(${TARGET} "${TARGET}ClangFormat")
         endif()
-        # If apply code style is on, turn off check code style
-        set(CHECK_CPP_CODE_STYLE OFF)
-        add_custom_target("${TARGET}ClangFormat"
-                COMMAND ${CLANG_FORMAT} -i ${TARGET_SRC_FILES}
-                WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}
-                COMMENT "Run clang format for ${TARGET}"
-                COMMAND_EXPAND_LISTS)
-        add_dependencies(${TARGET} "${TARGET}ClangFormat")
-    endif()
-    if(${CHECK_CPP_CODE_STYLE})
-        # Call the script to check formatting
-        add_custom_target("${TARGET}CodeStyleCheck"
-                COMMAND "${PYTHON}" ${CMAKE_SOURCE_DIR}/scripts/check_code_style.py
-                --sourcefile ${TARGET_SRC_FILES} --base ${CHECK_CPP_CODE_STYLE_BASE}
-                WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}
-                COMMENT "Check code style for ${TARGET}")
-        add_dependencies(${TARGET} "${TARGET}CodeStyleCheck")
+        if(${CHECK_CPP_CODE_STYLE})
+            # Call the script to check formatting
+            add_custom_target("${TARGET}CodeStyleCheck"
+                    COMMAND "${PYTHON}" ${CMAKE_SOURCE_DIR}/scripts/check_code_style.py
+                    --sourcefile ${TARGET_SRC_FILES} --base ${CHECK_CPP_CODE_STYLE_BASE}
+                    WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}
+                    COMMENT "Check code style for ${TARGET}")
+            add_dependencies(${TARGET} "${TARGET}CodeStyleCheck")
+        endif()
     endif()
 endfunction()

--- a/cmake/CodeStyle.cmake
+++ b/cmake/CodeStyle.cmake
@@ -17,8 +17,6 @@
 # Description: CMake code style build directives
 ###############################################################################
 
-cmake_minimum_required(VERSION 3.4.1)
-
 option(APPLY_CPP_CODE_STYLE "Apply C++ code style using clang format" OFF)
 option(CHECK_CPP_CODE_STYLE "Check C++ code style using clang format" OFF)
 option(CHECK_CPP_CODE_STYLE_BASE "Git branch/commit for C++ code style comparison" "HEAD")

